modules/auxiliary/admin/citrix/citrix_netscaler_config_decrypt.rb

```diff
@@ -0,0 +1,259 @@
+      OptPath.new('NS_KEK_F2', [ false, 'Path to NetScaler KEK fragment file F2.key' ]),
+      OptString.new('NS_IP', [ false, '(Optional) IPv4 address to attach to loot' ])
+    ])
+  end
+
+  def loot_host
+    datastore['NS_IP'] || '127.0.0.1'
+  end
+
+  def ns_conf
+    datastore['NS_CONF']
+  end
+
+  def ns_kek_f1
+    datastore['NS_KEK_F1']
+  end
+
+  def ns_kek_f2
+    datastore['NS_KEK_F2']
+  end
+
+  # ns.conf elements that contain potential secrets, update as needed
+  # k = parameter that has the secret (-key, -password, [...])
+  # v = start of config line that potentially has a secret
+  def ns_secret
+    {
+      'key' => ['add ssl certKey'],
+      'keyValue' => ['set ns encryptionParams'],
+      'radKey' => ['add authentication radiusAction'],
+      'ldapBindDnPassword' => ['add authentication ldapAction'],
+      'password' => ['set ns rpcNode', 'add lb monitor', 'add aaa user'],
+      'passPhrase' => ['add authentication dfaAction']
+    }
+  end
+
+  # Statically defined in libnscli90.so, modern appliances keep these in /nsconfig/.skf
+  def ns90_rc4key
+    '2286da6ca015bcd9b7259753c2a5fbc2'.scan(/../).map(&:hex).pack('C*')
+  end
+
+  def ns90_aeskey
+    '351cbe38f041320f22d990ad8365889c7de2fcccae5a1a8707e21e4adccd4ad9'.scan(/../).map(&:hex).pack('C*')
+  end
+
+  def run
+    if ns_kek_f1 && ns_kek_f2
+      print_status('Building NetScaler KEK from key fragments ...')
+      unless build_ns_kek
+        fail_with(Msf::Exploit::Failure::NoTarget, 'Unable to build NetScaler KEK from key fragments')
+      end
+    end
+    parse_ns_config
+  end
+
+  def build_ns_kek
+    unless File.size(ns_kek_f1) == 256 && File.size(ns_kek_f2) == 256
+      print_error('KEK files must be 256 bytes in size')
+      return false
+    end
+    f1_hex = File.read(ns_kek_f1)
+    f2_hex = File.read(ns_kek_f2)
+    f1_hex&.encode('UTF-8', 'binary')
+    f2_hex&.encode('UTF-8', 'binary')
+    unless f1_hex.match?(/^[0-9a-f]+$/i) && f2_hex.match?(/^[0-9a-f]+$/i)
+      print_error('NS KEK files must be hexadecimal format')
+      return false
+    end
+    f1_key = f1_hex[66..130].scan(/../).map(&:hex).pack('C*')
+    f2_key = f2_hex[70..134].scan(/../).map(&:hex).pack('C*')
+    f1_key_hex = f1_key.unpack('H*').first
+    f2_key_hex = f2_key.unpack('H*').first
+    print_good('NS KEK F1')
+    print_good("\t HEX: #{f1_key_hex}")
+    print_good('NS KEK F2')
+    print_good("\t HEX: #{f2_key_hex}")
+    @ns_kek_key = OpenSSL::HMAC.hexdigest('SHA256', f2_key, f1_key).scan(/../).map(&:hex).pack('C*')
+    @ns_kek_key_hex = @ns_kek_key.unpack('H*').first
+    print_good('Assembled NS KEK AES key')
+    print_good("\t HEX: #{@ns_kek_key_hex}\n")
+    true
+  rescue Encoding::UndefinedConversionError
+    print_error('Invalid NS KEK files provided: invalid UTF-8 data')
+    return false
+  rescue StandardError => e
+    print_error("#{__method__}: #{e.message}")
+    return false
```

Comment:
This is pretty liberal. Are there exceptions you're often running into that would take this path? It's going to end up being normalized to `fail_with(Failure::NoTarget` which may not be representative of the error that occurred. You're also not logging or printing the stack trace which will make it more difficult to figure out what actually happened if the issue needs to be resolved.